File: checker.py
#!/usr/bin/env python3

import bashlex
import untangle
import os
import re
import sys
import melodic_default


class MissingDependency:
    def __init__(self, dependency, position, line_number=None):
        self.dependency = dependency
        self.position = position
        self.line_number = line_number


def setup(path):
    grep_scripts = 'grep -rl ' + path + r' -e "#\(\!\)\{0,1\}/bin/bash"'
    grep_shebang = 'grep -rl ' + path + \
        r' -e "#\(\!\)\{0,1\}/bin\|usr\|sbin\|bash"'
    packages_pattern = 'find ' + path + r' -type f -name "package.xml"'
    try:
        scripts = list(
            map(lambda l: l.strip(), os.popen(grep_scripts).readlines()))
        shebangs = list(
            map(lambda l: l.strip(), os.popen(grep_shebang).readlines()))

        print(f'Found {len(scripts)} bash scripts.')
        print(f'Found {len(shebangs)} files containing shebangs (#!).')
        packages_file = os.popen(packages_pattern).readlines()[0].strip()
        parsed = untangle.parse(packages_file)

        if not hasattr(parsed, 'package') or (not hasattr(parsed.package, 'run_depend') and not hasattr(parsed.package, 'exec_depend')):
            return (scripts, shebangs, melodic_default)

        run_deps = set()
        exec_deps = set()
        if hasattr(parsed.package, 'run_depend'):
            run_deps = set(map(lambda e: e.cdata, parsed.package.run_depend))
        if hasattr(parsed.package, 'exec_depend'):
            exec_deps = set(map(lambda e: e.cdata, parsed.package.exec_depend))

        all_deps = melodic_default.all_binaries.union(
            run_deps).union(exec_deps)
        return (scripts, shebangs, all_deps)
    except Exception:
        return ([], [], [])


def validate(scripts, shebangs, deps):
    def validate_script(script):
        with open(script) as f:
            lines = f.readlines()
            if not lines or not '/bin/bash' in lines[0]:
                return None

            errors = []
            for i, line in enumerate(lines):
                result = validate_line(line)
                if not result:
                    continue
                result.line_number = i
                errors.append(result)
            if not errors:
                return None
            return (script, errors)

    def validate_shebang(shebang):
        try:
            with open(shebang) as f:
                first = f.readlines()[:1][0].strip()
                regexp = re.compile('(#!)+([/ ][a-z]+)+')
                command = regexp.findall(first)[0]
                command = command[1:][0][1:]
                if command not in deps:
                    return (shebang, [MissingDependency(command, (first.rfind(command), len(first)), 1)])
                return None
        except Exception:
            return None

    def validate_line(line):
        try:
            ast = bashlex.parsesingle(line.strip())
            error = dependency_exists(ast)
            return error
        except Exception:
            return None

    def dependency_exists(ast):
        if not ast or not ast.parts:
            return None

        command = ast.parts[:1][0]
        if command and command.kind is 'word' and command.word not in deps:
            return MissingDependency(command.word, command.pos)
        return None

    bash_missing = dict(
        filter(lambda r: r is not None,
               map(lambda s: validate_script(s), scripts)))

    shebang_missing = dict(
        filter(lambda r: r is not None,
               map(lambda s: validate_shebang(s), shebangs)))

    return (bash_missing, shebang_missing)


def print_result(result):
    if result:
        for script, errors in result.items():
            print('in \'{}\''.format(script))
            if not errors:
                print('    \t\t\t\tNone')
            if (errors):
                for e in errors:
                    print('    ln: {} {} \t\t\t{}'.format(
                        e.line_number, e.position, e.dependency))


if len(sys.argv) < 2:
    sys.exit('Please run the script with a path to a valid ROS package src.')

scripts, shebangs, run_deps = setup(sys.argv[1])
bash_scripts, shebang_scripts = validate(scripts, shebangs, run_deps)
print(f'Found {len(bash_scripts)} possible bash script errors.')
print(f'Found {len(shebang_scripts)} possible shebang (#!) errors.')
print_result(bash_scripts)
print_result(shebang_scripts)

File: src/checker.ts
#!/usr/bin/env node

import { readFileSync } from 'fs';
import parse = require('bash-parser');
import { defaults } from './melodic_default';
import { parse as parseXML } from 'fast-xml-parser';
import { shellSync } from 'execa';

export class MissingDependency {
  dependency: string;
  line: string;
  constructor(dependency: string, line: string) {
    this.dependency = dependency;
    const l = ` ${line}`;
    this.line = l.length > 40 ? `${l.substring(0, 40)}...` : l;
  }
}

const setup = (path: string) => {
  const grep_scripts = `grep -rl ${path} -e "#\\(\\!\\)\\{0,1\\}/bin/\\(bash\\|sh\\)"`;
  const packages_pattern = `find ${path} -type f -name "package.xml"`;
  const scripts = shellSync(grep_scripts).stdout.split('\n');
  const packages_files = shellSync(packages_pattern).stdout;
  const xmls = packages_files.split('\n').filter((l) => l !== '');
  let deps = new Set<string>();
  xmls.forEach((x) => {
    const xml = readFileSync(x);
    const parsed = parseXML(xml.toString());
    if (parsed['package']['run_depend']) {
      deps.add(parsed['package']['run_depend']);
    }
    if (parsed['package']['exec_depend']) {
      deps.add(parsed['package']['exec_depend']);
    }
  });
  const all_deps = new Set([...defaults, ...deps]);
  return { scripts, deps: all_deps };
};

const validate = (scripts: string[], deps: Set<string>) => {
  const validate_script = (script: string) => {
    const contents = readFileSync(script).toString();
    try {
      if (
        !contents ||
        (!contents.split('\n')[0].includes('bin/bash') &&
          !contents.split('\n')[0].includes('bin/sh'))
      ) {
        return { script, result: undefined };
      }

      const result = validate(contents);
      return { script, result };
    } catch (e) {
      return { script, result: [] };
    }
  };

  const validate = (contents) => {
    try {
      const ast = parse(contents);
      const error = dependency_exists(ast);
      return error;
    } catch (e) {
      return [];
    }
  };

  const dependency_exists = (ast) => {
    if (!ast) {
      return [];
    }

    const commands = ast.commands.filter((c) => c.type && c.type === 'Command');
    const errors: MissingDependency[] = [];
    commands.forEach((c) => {
      if (c.name) {
        const text = c.name.text;
        if (
          text &&
          !deps.has(text) &&
          text !== '/usr/bin/perl' &&
          !text.includes('./') &&
          !text.includes('$')
        ) {
          if (c.suffix) {
            errors.push(
              new MissingDependency(
                text,
                c.suffix.map((s) => s.text).join(' '),
              ),
            );
          } else errors.push(new MissingDependency(text, ''));
        }
      }
    });
    return errors;
  };

  const validated = scripts.map((s) => validate_script(s));
  const actualScripts = validated.filter((s) => s.result !== undefined);
  console.log(`Found ${actualScripts.length} bash scripts.`);
  const errors = validated.filter((o) => o && o.result && o.result.length);
  return errors;
};

const print_result = (result) => {
  if (!result) return;
  result.forEach((r) => {
    if (!r.result) return;
    console.log(`in '${r.script}'`);
    r.result.forEach((res) =>
      console.log(`\t\t\x1b[4m${res.dependency}\x1b[0m${res.line}`),
    );
  });
};

if (process.argv.length < 2) {
  console.log('Please run the script with a path to a valid ROS package src.');
  process.exit(1);
}

const scriptsAndDeps = setup(process.argv[2]);
const scripts = scriptsAndDeps.scripts;
const deps = scriptsAndDeps.deps;
const bash_scripts = validate(scripts, deps);
console.log(`Found ${bash_scripts.length} possible bash script errors.`);
print_result(bash_scripts);
